### DISCUSSION TITLE
Assign with referenced objects

### DISCUSSION THREAD

**Discussion by jeroni** [mikro-orm/mikro-orm#2920] [Ideas]
When using assign function with an object (usually from client side post / put actions) there's no way to reference objects when no id is present (new objects).

As a real example I have a `transport company` objects that has 2 collections: `vehicles` and `drivers`, but a `vehicle` object can have one of the `driver` objects assigned to itself. And imagine any of the objects has any id, they are all new.

As far as I'm concerned there's no way to reference that driver indicating it's really the same object placed in two levels:

```ts
// Transport company object
{
  name: 'Transport company 1',
  drivers: [ { name: 'd1' }, {name: 'd2' }],
  vehicles: [ { name: 'v1' }, {name: 'v2', driver: { name: 'd1' } ]
}
```
We use a generic function that handles this using negative id's from client side and doing the same process as assign but with this particularity.

We moved from objectionjs to Mikro a while ago and the previous lib had a way to do this by adding `#id: number` to one of the objects and `#ref: number` to indicate it had to be treated as the same object.

What do you think?

**Comment by B4nan** [comment]
Why dont you use client side generated PKs (e.g. uuid)? They are designed exactly for this use case. I don't really like that idea with `#id` and `#ref`, because it would requite quite a lot of changes type-wise (made them even more complex, which they already are) - hard to compare to objection that is not type safe at all (maybe I am wrong, never used it, but the docs look like its full of untyped strings everywhere). 

You would need to add those `#id/#ref` things yourself somewhere, where would that happen? If you build the entity instances there instead, it could already work - if there is an actual entity instance in the data, it will be used as is.

**Comment by jeroni** [comment]
I have always been concerned about uuid collision (very remote, I know). In that case, Mikro would perform an update instead of an insert operation?. And makes me ask if Mikro has to do a double operation, first check if the uuid exists on the table to decide which operation must be done? I believe the tree that holds the id indexing would be slower than with common serialised numbers.
I totally agree that #id/#ref into the objects is not the way to go in Mikro and the uuid seems the best solution. If some idea comes around I will make a proposal, something that gives more real-time control on the operations performed.

**Reply by B4nan** [reply]
If you want update queries, you _need_ to load the entities upfront to the context, otherwise there will be insert queries. 

Don't expect `assign` to magically solve your needs, most people think it will somehow do what they _think_ without providing any context to the ORM. `assign` is sync operation, it can't load anything, and the ORM/UoW can issue update queries only based on change sets - so you literally need to first load the entity, modify its state, and flush. Only way to make `assign` see a relation as a FK of sometihng already existing in the database is to pass the naked FK instead of an object (e.g. `assign({ author: 123 })` instead of `assign({ author: { id: 123 } })`).

**Reply by jeroni** [reply]
I see, thanks for the explanation!
